src/main/cpp/src/multiply.cu

```diff
@@ -16,99 +16,314 @@
 
 #include "multiply.hpp"
 #include "row_error_utilities.hpp"
+#include "utilities.hpp"
 
-#include <cudf/column/column_device_view.cuh>
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/valid_if.cuh>
 
 #include <rmm/device_uvector.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <cuda/std/limits>
 #include <thrust/for_each.h>
 
 namespace spark_rapids_jni {
 
 namespace {
 
-// Functor to perform multiplication on two columns with overflow checks if ANSI mode is enabled.
+void check_multiply_inputs(cudf::data_type left_type,
+                           cudf::data_type right_type,
+                           cudf::size_type left_size,
+                           cudf::size_type right_size,
+                           bool is_ansi_mode,
+                           bool is_try_mode)
+{
+  CUDF_EXPECTS(left_type == right_type, "Input columns must have the same data type");
+  CUDF_EXPECTS(is_basic_spark_numeric(left_type), "Unsupported data type for multiplication.");
+  CUDF_EXPECTS(left_size == right_size, "Input columns must have the same size");
+  CUDF_EXPECTS(!(is_ansi_mode && is_try_mode),
+               "Cannot enable both ANSI mode and TRY mode at the same time");
+}
+
+/**
+ * @brief Multiply two int8_t or int16_t values with overflow check.
+ * Set `valid` to false if overflow occurs.
+ */
+template <typename T, CUDF_ENABLE_IF(std::is_same_v<T, int8_t> || std::is_same_v<T, int16_t>)>
+__device__ void multiply(T x, T y, bool check_overflow, bool* valid, T* result)
+{
+  // when int8_t * int8_t or int16_t * int16_t, the result is promoted to int type.
+  int32_t r = x * y;
+  if (check_overflow &&
+      (r < cuda::std::numeric_limits<T>::min() || r > cuda::std::numeric_limits<T>::max())) {
+    *valid = false;  // overflow in ansi mode
+  } else {
+    *valid  = true;
+    *result = static_cast<T>(r);
+  }
+}
+
+/**
+ * @brief Multiply two int32_t values with overflow check.
+ * Set `valid` to false if overflow occurs.
+ */
+template <typename T, CUDF_ENABLE_IF(std::is_same_v<T, int32_t>)>
+__device__ void multiply(T x, T y, bool check_overflow, bool* valid, T* result)
+{
+  if (check_overflow) {
+    int64_t r = static_cast<int64_t>(x) * static_cast<int64_t>(y);
+    if (auto r_cast = static_cast<int>(r); static_cast<long>(r_cast) == r) {
+      *valid  = true;
+      *result = r_cast;
+    } else {
+      *valid = false;  // overflow in ansi mode
+    }
+  } else {
+    *valid  = true;
+    *result = x * y;
+  }
+}
+
+__device__ bool is_multiply_overflow(int64_t a, int64_t b)
+{
+  // constexpr int64_t INT64_MIN = (-9223372036854775807L - 1L); // -2^63
+  // constexpr int64_t INT64_MAX = 9223372036854775807L; // 2^63 - 1
+
+  if (a > 0 && b > 0 && a > 9223372036854775807L / b) return true;          // Positive overflow
+  if (a < 0 && b < 0 && a < 9223372036854775807L / b) return true;          // Negative overflow
+  if (a > 0 && b < 0 && b < (-9223372036854775807L - 1L) / a) return true;  // Mixed sign overflow
```

Comment:
Directly using INT64_MAX is causing building error, so using `-9223372036854775807L - 1L` is a work around.

Reply:
2^63 is not a prime. So is it possible that a * b == -2^63 but the overflow check fails?

Reply:
> if (a > 0 && b < 0 && b < (-9223372036854775807L - 1L) / a) return true
2^63 is not a prime. So is it possible that a * b == -2^63 but the overflow check fails?

I tried the following, the issue you described does not exist:
```cpp
#include <cstdio>

int main() {

  long a = 2L;
  long b = -4611686018427387904L;
  long min = -9223372036854775807L - 1L;

  if (a * b == min) {
    printf("a * b == min\n");
  }

  if (a > 0 && b < 0 && b < min / a ) {
    printf("overflow\n");
  } else {
    printf("no overflow\n");
  }
}

prints:

a * b == min
no overflow

```

Reply:
> I tried the following, the issue you described does not exist:

Ohh sorry, yes, you are right.